src/utilities/EventEmitter.test.ts

```diff
@@ -0,0 +1,242 @@
+    removeListener(targetListeners, listener);
+    expect(targetListeners).toStrictEqual([[altListener, altListener], [altListener]]);
+  });
+  it('Successfully removes a listener from objects', () => {
+    const targetListeners = [{
+      'fixture1': [...targetListenerArrayFixture1],
+      'fixture2': [...targetListenerArrayFixture2],
+    }];
+    removeListener(targetListeners, listener);
+    expect(targetListeners).toStrictEqual([
+      {
+        'fixture1': [altListener, altListener],
+        'fixture2': [altListener]
+      }
+    ]);
+  });
+
+  it('Successfully handles a mixture of objects and arrays', () => {
+    const targetListeners = [{
+      'fixture1': [...targetListenerArrayFixture1],
+      'fixture2': [...targetListenerArrayFixture2],
+    }, [...targetListenerArrayFixture1], [...targetListenerArrayFixture2]];
+    removeListener(targetListeners, listener);
+    expect(targetListeners).toStrictEqual([
+      {
+        'fixture1': [altListener, altListener],
+        'fixture2': [altListener]
+      },
+      [altListener, altListener], [altListener]
+    ],)
+  });
+
+  it('Only removes listeners from a filtered list of events', () => {
+    const targetListeners = [{
+      'fixture1': [...targetListenerArrayFixture1],
+      'fixture2': [...targetListenerArrayFixture2],
+    }];
+    removeListener(targetListeners, listener, 'fixture2');
+    expect(targetListeners).toStrictEqual([
+      {
+        'fixture1': [altListener, listener, altListener],
+        'fixture2': [altListener]
+      }
+    ]);
+    removeListener(targetListeners, altListener, 'fixture1');
+    expect(targetListeners).toStrictEqual([
+      {
+        'fixture1': [listener],
+        'fixture2': [altListener]
+      }
+    ]);
+  })
+});
+
+describe('Event emitter class', () => {
+  let eventEmitter: EventEmitter<{ 'myEvent': string, 'myOtherEvent': string, 'myThirdEvent': string }>;
+  const listener = () => true;
+  const altListener = () => false;
+
+  afterEach(() => {
+    eventEmitter = new EventEmitter();
+  });
+
+  describe('calling the on method', () => {
+    beforeEach(() => {
+      eventEmitter = new EventEmitter();
+    });
+    it('Adds a listener to the "any" set of event listeners', () => {
+      eventEmitter.on(listener);
+      expect(eventEmitter.any).toStrictEqual([listener]);
+    });
+    it('Adds a listener to a provided field of an event listener', () => {
+      eventEmitter.on('myEvent', listener);
+      expect(eventEmitter.events['myEvent']).toStrictEqual([listener]);
+    });
+    it('Adds a listener to all provided fields of an event listener', () => {
+      eventEmitter.on(['myEvent', 'myOtherEvent', 'myThirdEvent'], listener);
+      expect(eventEmitter.events['myEvent']).toStrictEqual([listener]);
+      expect(eventEmitter.events['myOtherEvent']).toStrictEqual([listener]);
+      expect(eventEmitter.events['myThirdEvent']).toStrictEqual([listener]);
```

Comment:
Should we not test here `.on` API via calling `.emit` instead of only looking at `.events`, which technically is just an implementation detail? None of these tests, as far as I can tell, test if the callback is actually called?

Reply:
As an example - if I change the code so that the listener in `.on` is no longer called, all these tests would still pass (or am I missing something here @MymmiJ) ?

Reply:
I see events as more than an implementation detail I suppose; `on` is not for me a function that necessarily queues an event to be `emit`ted, it should have an ability to record events to execute even if `emit` doesn't exist or there is no way to emit them. Also, it shouldn't only test that `emit` works on an event, because we want to know that the right quantity of listeners have been registered in the correct places.

An example of where this idea matters already is in `once`, where we can accumulate events by calling `once(['myEvent', 'myOtherEvent'], listener);` and then calling `eventEmitter.emit(listener);`, which will leave listener wrapping functions in place. As far as I can tell, this seems to be hard to avoid and it's unlikely to be a common pattern, so I've not tried to fix; but it's an example of a case where calling a method to register an event and calling a method to emit an event results in the listener being called, but the result is still wrong.

Reply:
Ideally, we would test exactly how a user uses the API, so we know if we break features.

I think it's ok to do additional "whitebox testing" and reach into the implementation details (appreciating that these tests will be brittle because they know about details like the `.events`), but I disagree that this should be the default and the actual usage tests are skipped completely.

As an example, if I only see the tests, I need to read the whole file to understand that I need to use both `.emit` and `.on` to use the class. Tests do double duty as documentation, so I think this is an important aspect that needs to considered here.

I'll have another look at the `.once`, if there is a bug we copied over (or I've introduced in the refactor) we should fix it.

>I think these tests should still pass if you change the code so that the listener in on is not called, but the tests in emit should not pass (because registering events works, but emitting events is broken).

But I think this is wrong, `.on` should break. The promise of `.on` is that it will call the listener, if that method is broken why should `.emit` break (from the user's perspective - let's be ignorant about implementation details here).

Reply:
>  Tests do double duty as documentation, so I think this is an important aspect that needs to considered here.

I don't know if our current tests really follow that philosophy, considering the majority of space tests directly inject presence events rather than using space.enter()

Reply:
@Peter-Maguire I would say they do. The tests listen for events & test the specifics of how the handler was called. They don't check for internal methods. The mocking done for space.enter() is just the test setup, it's there, so the test can be more concise. We could use `space.enter` there, mock it, and it would change nothing about the test except lengthening it. 

Here, we depend on the implementation detail for the test result. As mentioned above, there is a case for that, but we should try to use the method as a user would first.

Reply:
That's all good with me, I will rewrite these to test on + emit and once + emit together